File: ml-pipeline/indicators.py
"""
Technical indicators -- the Python side of a two-language pair.

This is a deliberate mirror of backend/src/services/indicators.js. The UI
panel renders the Node values while the model trains on these; if they
drift, the site contradicts its own model. tests/test_indicators.py checks
both against a shared fixture so drift fails CI rather than reaching users.

Conventions (identical to the JS):
  * RSI-14 and ATR-14 use Wilder smoothing (alpha = 1/n), seeded with the
    SMA of the first n values -- NOT a rolling mean. This is the usual
    source of mismatch with charting platforms.
  * MACD = EMA(12) - EMA(26); signal = EMA(9) of the MACD line, computed
    over the defined portion only.
  * EMA is seeded with the SMA of the first `period` values.
  * Outputs are float arrays the same length as the input, padded with NaN
    where undefined -- never 0, which a model would read as a real value.
"""

from __future__ import annotations

import numpy as np
import pandas as pd


def _as_array(values) -> np.ndarray:
    return np.asarray(values, dtype=float)


def sma(values, period: int) -> np.ndarray:
    v = _as_array(values)
    out = np.full(len(v), np.nan)
    if period <= 0 or len(v) < period:
        return out
    cumsum = np.cumsum(np.insert(v, 0, 0.0))
    out[period - 1:] = (cumsum[period:] - cumsum[:-period]) / period
    return out


def ema(values, period: int) -> np.ndarray:
    """EMA seeded with the SMA of the first `period` values."""
    v = _as_array(values)
    out = np.full(len(v), np.nan)
    if period <= 0 or len(v) < period:
        return out
    k = 2.0 / (period + 1.0)
    prev = v[:period].mean()
    out[period - 1] = prev
    for i in range(period, len(v)):
        prev = v[i] * k + prev * (1.0 - k)
        out[i] = prev
    return out


def wilder_smooth(values, period: int) -> np.ndarray:
    """Wilder's RMA: alpha = 1/period, seeded with the SMA."""
    v = _as_array(values)
    out = np.full(len(v), np.nan)
    if period <= 0 or len(v) < period:
        return out
    prev = v[:period].mean()
    out[period - 1] = prev
    for i in range(period, len(v)):
        prev = (prev * (period - 1) + v[i]) / period
        out[i] = prev
    return out


def rsi(closes, period: int = 14) -> np.ndarray:
    c = _as_array(closes)
    out = np.full(len(c), np.nan)
    if len(c) <= period:
        return out

    diff = np.diff(c)
    gains = np.where(diff > 0, diff, 0.0)
    losses = np.where(diff < 0, -diff, 0.0)

    avg_gain = wilder_smooth(gains, period)
    avg_loss = wilder_smooth(losses, period)

    for i in range(len(gains)):
        if np.isnan(avg_gain[i]) or np.isnan(avg_loss[i]):
            continue
        if avg_loss[i] == 0:
            # No losses. A window with gains is genuinely maxed at 100, but
            # a completely FLAT window has no gains either -- that is
            # neutral. Mirrors the JS implementation.
            out[i + 1] = 50.0 if avg_gain[i] == 0 else 100.0
            continue
        out[i + 1] = 100.0 - 100.0 / (1.0 + avg_gain[i] / avg_loss[i])
    return out


def macd(closes, fast: int = 12, slow: int = 26, signal_period: int = 9):
    c = _as_array(closes)
    ema_fast = ema(c, fast)
    ema_slow = ema(c, slow)
    line = ema_fast - ema_slow  # NaN propagates where either is undefined

    signal = np.full(len(c), np.nan)
    histogram = np.full(len(c), np.nan)

    defined = np.where(~np.isnan(line))[0]
    if len(defined):
        start = defined[0]
        dense_signal = ema(line[start:], signal_period)
        signal[start:] = dense_signal
        histogram = line - signal
    return line, signal, histogram


def true_range(highs, lows, closes) -> np.ndarray:
    h, l, c = _as_array(highs), _as_array(lows), _as_array(closes)
    out = np.full(len(c), np.nan)
    if len(c) == 0:
        return out
    out[0] = h[0] - l[0]
    prev_close = c[:-1]
    out[1:] = np.maximum.reduce([
        h[1:] - l[1:],
        np.abs(h[1:] - prev_close),
        np.abs(l[1:] - prev_close),
    ])
    return out


def atr(highs, lows, closes, period: int = 14) -> np.ndarray:
    """
    Average True Range, Wilder-smoothed.

    Undefined true-range bars (a halted session with no high/low) are
    skipped rather than propagated. Feeding them straight into
    wilder_smooth would make a single gap poison every subsequent value --
    one missing bar turned the entire column NaN, which downstream would
    silently drop every training row. This also keeps the result identical
    to the Node implementation, which filters the same way.
    """
    tr = true_range(highs, lows, closes)
    defined = ~np.isnan(tr)
    if defined.sum() < period:
        return np.full(len(tr), np.nan)

    smoothed = wilder_smooth(tr[defined], period)
    out = np.full(len(tr), np.nan)
    out[defined] = smoothed
    return out


def bollinger(closes, period: int = 20, mult: float = 2.0):
    c = _as_array(closes)
    middle = sma(c, period)
    upper = np.full(len(c), np.nan)
    lower = np.full(len(c), np.nan)
    percent_b = np.full(len(c), np.nan)

    for i in range(period - 1, len(c)):
        window = c[i - period + 1: i + 1]
        # Population std (ddof=0), matching the JS implementation.
        sd = window.std(ddof=0)
        upper[i] = middle[i] + mult * sd
        lower[i] = middle[i] - mult * sd
        span = upper[i] - lower[i]
        if span:
            percent_b[i] = (c[i] - lower[i]) / span * 100.0
    return middle, upper, lower, percent_b


def roc(closes, period: int = 10) -> np.ndarray:
    c = _as_array(closes)
    out = np.full(len(c), np.nan)
    if len(c) > period:
        prior = c[:-period]
        with np.errstate(divide="ignore", invalid="ignore"):
            out[period:] = np.where(prior != 0, (c[period:] - prior) / prior * 100.0, np.nan)
    return out


def compute_all(df: pd.DataFrame) -> pd.DataFrame:
    """
    Add every indicator column to an OHLCV frame.

    Args:
        df: columns open/high/low/close/volume, chronological, oldest first.
    Returns:
        A copy with indicator columns appended.
    """
    out = df.copy()
    close = out["close"].to_numpy(dtype=float)
    high = out["high"].fillna(out["close"]).to_numpy(dtype=float)
    low = out["low"].fillna(out["close"]).to_numpy(dtype=float)

    out["sma20"] = sma(close, 20)
    out["sma50"] = sma(close, 50)
    out["sma200"] = sma(close, 200)
    out["ema20"] = ema(close, 20)
    out["ema50"] = ema(close, 50)
    out["ema200"] = ema(close, 200)
    out["rsi14"] = rsi(close, 14)
    out["atr14"] = atr(high, low, close, 14)

    macd_line, macd_signal, macd_hist = macd(close)
    out["macd"] = macd_line
    out["macd_signal"] = macd_signal
    out["macd_histogram"] = macd_hist

    bb_mid, bb_up, bb_low, bb_pct = bollinger(close)
    out["bb_middle"] = bb_mid
    out["bb_upper"] = bb_up
    out["bb_lower"] = bb_low
    out["bb_percent_b"] = bb_pct

    out["roc10"] = roc(close, 10)
    return out


__all__ = [
    "sma", "ema", "wilder_smooth", "rsi", "macd", "atr", "true_range",
    "bollinger", "roc", "compute_all",
]

File: backend/src/services/indicators.js
/**
 * Technical indicators.
 *
 * These serve two masters: the UI indicators panel, and (via the Python
 * pipeline) the LSTM's feature vector. The two MUST agree numerically or
 * the site will contradict its own model, so the parameters below are
 * fixed and mirrored exactly in ml-pipeline/indicators.py. A shared
 * fixture (tests/fixtures/indicators.json) is asserted by both sides.
 *
 * Conventions, chosen to match the papers and standard charting platforms:
 *  - RSI-14 and ATR-14 use **Wilder** smoothing (alpha = 1/n), not a
 *    simple mean. This is the single most common source of mismatch
 *    between naive implementations and TradingView.
 *  - MACD is EMA(12) - EMA(26), signal EMA(9) of the MACD line.
 *  - EMA seeds from the SMA of the first `period` values.
 *  - Output arrays are the same length as the input, left-padded with
 *    `null` where the indicator is not yet defined. Never 0 — a zero would
 *    be read as a real value by both the chart and the model.
 */

/** Simple moving average. */
export function sma(values, period) {
  const out = new Array(values.length).fill(null);
  if (period <= 0 || values.length < period) return out;
  let sum = 0;
  for (let i = 0; i < values.length; i += 1) {
    sum += values[i];
    if (i >= period) sum -= values[i - period];
    if (i >= period - 1) out[i] = sum / period;
  }
  return out;
}

/** Exponential moving average, seeded with the SMA of the first `period`. */
export function ema(values, period) {
  const out = new Array(values.length).fill(null);
  if (period <= 0 || values.length < period) return out;
  const k = 2 / (period + 1);
  let seed = 0;
  for (let i = 0; i < period; i += 1) seed += values[i];
  let prev = seed / period;
  out[period - 1] = prev;
  for (let i = period; i < values.length; i += 1) {
    prev = values[i] * k + prev * (1 - k);
    out[i] = prev;
  }
  return out;
}

/** Wilder's smoothing (RMA): alpha = 1/period, seeded with the SMA. */
export function wilderSmooth(values, period) {
  const out = new Array(values.length).fill(null);
  if (period <= 0 || values.length < period) return out;
  let seed = 0;
  for (let i = 0; i < period; i += 1) seed += values[i];
  let prev = seed / period;
  out[period - 1] = prev;
  for (let i = period; i < values.length; i += 1) {
    prev = (prev * (period - 1) + values[i]) / period;
    out[i] = prev;
  }
  return out;
}

/** Relative Strength Index, Wilder-smoothed. Range 0-100. */
export function rsi(closes, period = 14) {
  const out = new Array(closes.length).fill(null);
  if (closes.length <= period) return out;

  const gains = new Array(closes.length - 1);
  const losses = new Array(closes.length - 1);
  for (let i = 1; i < closes.length; i += 1) {
    const diff = closes[i] - closes[i - 1];
    gains[i - 1] = diff > 0 ? diff : 0;
    losses[i - 1] = diff < 0 ? -diff : 0;
  }

  const avgGain = wilderSmooth(gains, period);
  const avgLoss = wilderSmooth(losses, period);

  for (let i = 0; i < gains.length; i += 1) {
    if (avgGain[i] == null || avgLoss[i] == null) continue;
    if (avgLoss[i] === 0) {
      // No losses at all. A window with gains is genuinely maxed out at
      // 100, but a completely FLAT window has no gains either — that is
      // neutral, not overbought. Without this an illiquid stock that has
      // not moved for 14 sessions would be badged "Overbought".
      out[i + 1] = avgGain[i] === 0 ? 50 : 100;
      continue;
    }
    out[i + 1] = 100 - 100 / (1 + avgGain[i] / avgLoss[i]);
  }
  return out;
}

/** MACD(12, 26, 9). Returns the line, its signal, and the histogram. */
export function macd(closes, fast = 12, slow = 26, signalPeriod = 9) {
  const emaFast = ema(closes, fast);
  const emaSlow = ema(closes, slow);

  const line = closes.map((_, i) =>
    emaFast[i] != null && emaSlow[i] != null ? emaFast[i] - emaSlow[i] : null,
  );

  // The signal EMA runs over the defined portion of the MACD line only.
  const firstDefined = line.findIndex((v) => v != null);
  const signal = new Array(closes.length).fill(null);
  const histogram = new Array(closes.length).fill(null);

  if (firstDefined !== -1) {
    const dense = line.slice(firstDefined);
    const denseSignal = ema(dense, signalPeriod);
    for (let i = 0; i < dense.length; i += 1) {
      const idx = firstDefined + i;
      signal[idx] = denseSignal[i];
      if (denseSignal[i] != null) histogram[idx] = line[idx] - denseSignal[i];
    }
  }

  return { macd: line, signal, histogram };
}

/** True Range series. TR[0] is high-low (no prior close available). */
export function trueRange(highs, lows, closes) {
  const out = new Array(closes.length).fill(null);
  for (let i = 0; i < closes.length; i += 1) {
    if (highs[i] == null || lows[i] == null) continue;
    if (i === 0) { out[i] = highs[i] - lows[i]; continue; }
    const pc = closes[i - 1];
    out[i] = Math.max(highs[i] - lows[i], Math.abs(highs[i] - pc), Math.abs(lows[i] - pc));
  }
  return out;
}

/** Average True Range, Wilder-smoothed. */
export function atr(highs, lows, closes, period = 14) {
  const tr = trueRange(highs, lows, closes);
  const dense = tr.filter((v) => v != null);
  if (dense.length < period) return new Array(closes.length).fill(null);
  const smoothed = wilderSmooth(dense, period);
  // Re-align onto the original index space.
  const out = new Array(closes.length).fill(null);
  let j = 0;
  for (let i = 0; i < tr.length; i += 1) {
    if (tr[i] == null) continue;
    out[i] = smoothed[j];
    j += 1;
  }
  return out;
}

/** Bollinger Bands(20, 2). Uses the population standard deviation. */
export function bollinger(closes, period = 20, mult = 2) {
  const middle = sma(closes, period);
  const upper = new Array(closes.length).fill(null);
  const lower = new Array(closes.length).fill(null);
  const width = new Array(closes.length).fill(null);
  const percentB = new Array(closes.length).fill(null);

  for (let i = period - 1; i < closes.length; i += 1) {
    const mean = middle[i];
    if (mean == null) continue;
    let variance = 0;
    for (let j = i - period + 1; j <= i; j += 1) variance += (closes[j] - mean) ** 2;
    const sd = Math.sqrt(variance / period);
    upper[i] = mean + mult * sd;
    lower[i] = mean - mult * sd;
    width[i] = mean ? ((upper[i] - lower[i]) / mean) * 100 : null;
    const span = upper[i] - lower[i];
    percentB[i] = span ? ((closes[i] - lower[i]) / span) * 100 : null;
  }
  return { middle, upper, lower, width, percentB };
}

/** Rate of change over `period` bars, as a percentage. */
export function roc(closes, period = 10) {
  const out = new Array(closes.length).fill(null);
  for (let i = period; i < closes.length; i += 1) {
    if (closes[i - period]) out[i] = ((closes[i] - closes[i - period]) / closes[i - period]) * 100;
  }
  return out;
}

/**
 * Compute the full indicator set from OHLCV candles.
 * @param {{o,h,l,c,v}[]} candles chronological, oldest first
 * @returns {{series:object, latest:object}} full series plus the last value
 *          of each, which is what the UI panel renders.
 */
export function computeAll(candles) {
  const closes = candles.map((d) => d.c);
  const highs = candles.map((d) => d.h ?? d.c);
  const lows = candles.map((d) => d.l ?? d.c);

  const macdOut = macd(closes);
  const bb = bollinger(closes);

  const series = {
    sma20: sma(closes, 20),
    sma50: sma(closes, 50),
    sma200: sma(closes, 200),
    ema20: ema(closes, 20),
    ema50: ema(closes, 50),
    ema200: ema(closes, 200),
    rsi14: rsi(closes, 14),
    atr14: atr(highs, lows, closes, 14),
    macd: macdOut.macd,
    macdSignal: macdOut.signal,
    macdHistogram: macdOut.histogram,
    bbUpper: bb.upper,
    bbMiddle: bb.middle,
    bbLower: bb.lower,
    bbPercentB: bb.percentB,
    roc10: roc(closes, 10),
  };

  const last = (arr) => (arr.length ? arr[arr.length - 1] : null);
  const latest = Object.fromEntries(Object.entries(series).map(([k, v]) => [k, last(v)]));

  // Interpretations the UI shows as badges. Thresholds are the textbook ones.
  const price = last(closes);
  latest.price = price;
  latest.rsiSignal =
    latest.rsi14 == null ? null
      : latest.rsi14 >= 70 ? 'overbought'
      : latest.rsi14 <= 30 ? 'oversold'
      : 'neutral';
  latest.macdSignalDirection =
    latest.macd == null || latest.macdSignal == null ? null
      : latest.macd > latest.macdSignal ? 'bullish' : 'bearish';
  latest.trend =
    price == null || latest.sma50 == null || latest.sma200 == null ? null
      : price > latest.sma50 && latest.sma50 > latest.sma200 ? 'uptrend'
      : price < latest.sma50 && latest.sma50 < latest.sma200 ? 'downtrend'
      : 'sideways';

  return { series, latest };
}

export default {
  sma, ema, wilderSmooth, rsi, macd, atr, trueRange, bollinger, roc, computeAll,
};
